crates/sui/src/client_ptb/ptb.rs

```diff
@@ -149,85 +139,44 @@ impl PTB {
+            rest: Opts {
+                dry_run: program_metadata.dry_run_set,
+                gas_budget: program_metadata.gas_budget.map(|x| x.value),
+                serialize_unsigned_transaction: program_metadata.serialize_unsigned_set,
+                serialize_signed_transaction: program_metadata.serialize_signed_set,
+            },
         };
 
-        // find the gas coins if we have no gas coin given
-        let coins = if let Some(gas) = gas {
-            context.get_object_ref(gas).await?
-        } else {
-            context
-                .gas_for_owner_budget(sender, gas_budget, BTreeSet::new())
-                .await?
-                .1
-                .object_ref()
-        };
-
-        if program_metadata.dry_run_set {
-            let response = execute_dry_run(
-                context,
-                sender,
-                tx_kind,
-                Some(gas_budget),
-                gas_price,
-                Some(vec![coins.0]),
-                None,
-            )
-            .await?;
-            println!("{}", response);
-            return Ok(());
-        }
-
-        // create the transaction data that will be sent to the network
-        let tx_data = TransactionData::new(tx_kind.clone(), sender, coins, gas_budget, gas_price);
-
-        if program_metadata.serialize_unsigned_set {
-            serialize_or_execute!(tx_data, true, false, context, PTB).print(true);
-            return Ok(());
-        }
+        let transaction_response = dry_run_or_execute_or_serialize(
+            sender,
+            tx_kind,
+            context,
+            None,
+            None,
+            Some(opts),
+            None::<Opts>,
+        )
+        .await?;
 
-        if program_metadata.serialize_signed_set {
-            serialize_or_execute!(tx_data, false, true, context, PTB).print(true);
+        if let SuiClientCommandResult::DryRun(response) = transaction_response {
+            println!("{}", Pretty(&response));
             return Ok(());
         }
 
-        // sign the tx
-        let signature =
-            context
-                .config
-                .keystore
-                .sign_secure(&sender, &tx_data, Intent::sui_transaction())?;
-
-        // execute the transaction
-        let transaction_response = context
-            .get_client()
-            .await?
-            .quorum_driver_api()
-            .execute_transaction_block(
-                Transaction::from_data(tx_data, vec![signature]),
-                SuiTransactionBlockResponseOptions::full_content(),
-                Some(ExecuteTransactionRequestType::WaitForLocalExecution),
-            )
-            .await?;
+        let transaction_response =
+            if let SuiClientCommandResult::TransactionBlock(response) = transaction_response {
+                response
+            } else {
+                anyhow::bail!("Internal error. Cannot run the PTB")
```

Comment:
Is this always an internal error, or can we fail to get a response because of a user error?

Reply:
But of course, if the network fails during the tx, I am not sure what error will be returned. I assume some `network` related error, and not this one.